File: crates/bask-python/examples/gpu.py
# SPDX-FileCopyrightText: 2026 Wavelens GmbH <info@wavelens.io>
#
# SPDX-License-Identifier: MIT OR Apache-2.0
"""Resource-attribute-aware retry. Two GPU instances self-describe with a `gpu` attribute
and draw from a shared `gpu` pool; a job too large for the a100 raises different_attr("gpu")
and is retried on the h100, while a job with no data raises Fatal and lands in the sink.
"""
from bask import Engine, Retry, Fatal, different_attr


class TrainJob:
    def __init__(self, id, size_gb):
        self.id = id
        self.size_gb = size_gb


class Trainer:
    def __init__(self, kind, vram_gb):
        self.kind = kind
        self.vram_gb = vram_gb

    def process(self, job, ctx):
        if job.size_gb == 0:
            raise Fatal(f"job {job.id} has no data")
        if job.size_gb > self.vram_gb:
            raise different_attr("gpu", f"job {job.id} needs {job.size_gb}GB, {self.kind} has {self.vram_gb}GB")
        ctx.route(Placed, (job.id, self.kind))


dead = []


def on_dead(letter):
    dead.append(letter["task"].id)
    print(f"dead-letter: {letter['error']} ({letter['attempts']} attempts)")


engine = Engine(
    concurrency=1,
    retry=Retry(max_attempts=2, jitter=0.2),
    resources={"gpu": 2},
    dead_letter=on_dead,
)


@engine.router
class Placed:
    def __init__(self):
        self.hits = []

    def route(self, hit, out):
        self.hits.append(hit)

    def finalize(self):
        return sorted(self.hits)


engine.register(TrainJob, Trainer("a100", 40), label="a100", attrs={"gpu": "a100"}, requires=["gpu"])
engine.register(TrainJob, Trainer("h100", 80), label="h100", attrs={"gpu": "h100"}, requires=["gpu"])

for job_id, size_gb in [(1, 20), (2, 60), (3, 0), (4, 75), (5, 35)]:
    engine.seed(TrainJob(job_id, size_gb))
report = engine.run()

print("\nplacements:")
for job_id, gpu in report.output(Placed):
    print(f"  job {job_id} -> {gpu}")
print("\ndead-lettered jobs:", dead)
print("stats:", report)

File: crates/bask/examples/gpu.rs
/*
 * SPDX-FileCopyrightText: 2026 Wavelens GmbH <info@wavelens.io>
 *
 * SPDX-License-Identifier: MIT OR Apache-2.0
 */

//! Resource-attribute-aware retry. Two GPU instances self-describe with a `gpu`
//! attribute and draw from a shared `gpu` pool; a job too large for the a100 fails with a
//! `DifferentAttr("gpu")` hint and is retried on the h100, while a job with no data fails
//! fatally and lands in the dead-letter sink.
use std::sync::{Arc, Mutex};

use bask::prelude::*;

struct TrainJob {
    id: u32,
    size_gb: u32,
}

/// Records the gpu each job was ultimately placed on.
struct Placed;
impl Router for Placed {
    type Input = (u32, &'static str);
    type State = Vec<(u32, &'static str)>;
    type Output = Vec<(u32, &'static str)>;
    fn route(state: &mut Self::State, hit: (u32, &'static str), _out: &mut Emit) {
        state.push(hit);
    }
    fn merge(left: &mut Self::State, right: Self::State) {
        left.extend(right);
    }
    fn finalize(mut state: Self::State) -> Self::Output {
        state.sort();
        state
    }
}

struct Trainer {
    kind: &'static str,
    vram_gb: u32,
}
#[async_trait]
impl Worker for Trainer {
    type Task = TrainJob;
    async fn process(&self, job: &TrainJob, ctx: &Context) -> anyhow::Result<()> {
        if job.size_gb == 0 {
            return Err(anyhow::anyhow!("job {} has no data", job.id)).fatal();
        }
        if job.size_gb > self.vram_gb {
            return Err(anyhow::anyhow!(
                "job {} needs {}GB, {} has {}GB",
                job.id,
                job.size_gb,
                self.kind,
                self.vram_gb
            ))
            .retry_on(RetryOn::DifferentAttr("gpu".into()));
        }
        ctx.route::<Placed>((job.id, self.kind)).await?;
        Ok(())
    }
}

#[tokio::main]
async fn main() -> anyhow::Result<()> {
    let dead: Arc<Mutex<Vec<u32>>> = Arc::new(Mutex::new(Vec::new()));
    let sink = {
        let dead = dead.clone();
        move |letter: DeadLetter| {
            if let Some(job) = letter.payload.downcast_ref::<TrainJob>() {
                dead.lock().unwrap().push(job.id);
            }
            println!(
                "dead-letter: {} ({} attempts)",
                letter.error, letter.attempts
            );
        }
    };

    let mut builder = Engine::builder()
        .resource("gpu", 2)
        .worker_cfg(
            Trainer {
                kind: "a100",
                vram_gb: 40,
            },
            WorkerCfg::new()
                .label("a100")
                .attr("gpu", "a100")
                .requires("gpu"),
        )
        .worker_cfg(
            Trainer {
                kind: "h100",
                vram_gb: 80,
            },
            WorkerCfg::new()
                .label("h100")
                .attr("gpu", "h100")
                .requires("gpu"),
        )
        .router::<Placed>()
        .retry(RetryPolicy::new().max_attempts(2).jitter(0.2))
        .dead_letter(sink)
        .concurrency(1);

    let jobs = [(1, 20), (2, 60), (3, 0), (4, 75), (5, 35)];
    for (id, size_gb) in jobs {
        builder = builder.seed(TrainJob { id, size_gb });
    }
    let report = builder.run().await?;

    println!("\nplacements:");
    for (id, gpu) in report.output::<Placed>().unwrap() {
        println!("  job {id} -> {gpu}");
    }
    println!("\ndead-lettered jobs: {:?}", dead.lock().unwrap());
    println!("stats: {:?}", report.stats);
    Ok(())
}
